src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,268 @@
+#include <shogun/machine/IterativeMachine.h>
+#include <shogun/mathematics/RandomMixin.h>
+#include <shogun/distributions/Distribution.h>
+#include <shogun/optimization/GradientDescendUpdater.h>
+#include <shogun/optimization/ConstLearningRate.h>
+#include <shogun/optimization/ElasticNetPenalty.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+#include <shogun/optimization/FirstOrderCostFunction.h>
+
+namespace shogun
+{
+    enum GLM_DISTRIBUTION
+	{
+        //TODO GAUSSIAN,
+        //TODO BINOMIAL,
+        //TODO GAMMA,
+        //TODO SOFTPLUS,
+        //TODO PROBIT,
+        POISSON
+	};
+
+class DotFeatures;
+class Features;
+class RegressionLabels;
+class GLMCostFunction;
+
+/** @brief Class for estimating regularized generalized linear models (GLM).
+ *   The regularized GLM minimizes the penalized negative log likelihood:
+ *
+ *  This uses Elastic-net penalty which defaults to the ridge penalty when
+ *  alpha = 0 and defaults to the lasso penalty when alpha = 1.
+ *
+ * */
+class GLM : public RandomMixin<IterativeMachine<LinearMachine>>
+{
+ 	public:
+		friend class GLMCostFunction;
+		
+		/** default constructor */
+		GLM();
+
+ 		GLM(GLM_DISTRIBUTION distribution, float64_t alpha, float64_t lambda, float64_t learning_rate, int32_t max_iterations, float64_t tolerance, float64_t eta);
+
+ 		/** destructor */
+ 		~GLM() override {}
+
+		/** apply linear machine to data
+		 * for regression problem
+		 *
+		 * @param data (test)data to be classified
+		 * @return classified labels
+		 */
+		std::shared_ptr<RegressionLabels> apply_regression(std::shared_ptr<Features> data=NULL) override;
+
+ 		/** Returns the name of the SGSerializable instance.  It MUST BE
+ 		 *  the CLASS NAME without the prefixed `C'.
+ 		 *
+ 		 * @return name of the SGSerializable
+ 		 */
+ 		const char* get_name() const override { return "GLM"; }
+
+		void set_tau(SGMatrix<float64_t> tau);
+
+		SGMatrix<float64_t> get_tau();
+
+	protected:
+
+		void init_model(const std::shared_ptr<Features>& data);
+		
+		void iteration() override;
+
+	protected:
+
+		/** Distribution type */
+		GLM_DISTRIBUTION distribution = POISSON;
+
+		/** a threshold parameter that linearizes the exp() function above eta. */
+		float64_t m_eta = 2.0;
+
+		/** regularization parameter :math:`\\lambda` of penalty term. */
+		float64_t m_lambda = 0.1;
+
+		/** the weighting between L1 penalty and L2 penalty term of the loss function. */
+		float64_t m_alpha = 0.5;
+
+		/** convergence threshold or stopping criteria. Optimization loop will stop when relative change in parameter norm is below the threshold. */
+		float64_t m_tolerance = 1e-6;
+
+		bool m_compute_bias = true;
+
+		std::shared_ptr<GradientDescendUpdater> m_gradient_updater;
+
+		std::shared_ptr<ConstLearningRate> m_learning_rate;
+
+		std::shared_ptr<ElasticNetPenalty> m_penalty;
+
+		std::shared_ptr<GLMCostFunction> m_cost_function;
+};
+
+class GLMCostFunction: public FirstOrderCostFunction
+{
+public:
```

Comment:
Missing constructor

Reply:
Do I need to explicitly define one? There is a default constructor which is called otherwise right.

Reply:
If you don't initialise then it will just point to some memory location. That memory location may be valid, but will be garbage because the object at that location has not been initialised. You have something like this https://godbolt.org/z/sLbM9o, and the return value is a result of undefined behaviour.

Reply:
I see that you have `set_target` but wouldn't it be possible to have constructor with `std::shared_ptr<GLM>&` argument instead? You should look into [RAII](https://en.cppreference.com/w/cpp/language/raii), as this is best practice.

Reply:
Hi, How will I implement this here though?
For there to be no segmentation fault, I'll need to call the constructor for `GLMCostFunction` in the constructor of `GLM`.
But I can't initialize `GLMCostFunction` in `GLM`'s constructor as I need to pass a `shared_pointer` to the `GLMCostFunction` which is not possible until the object is actually created.

https://stackoverflow.com/questions/31924396/why-shared-from-this-cant-be-used-in-constructor-from-technical-standpoint

Is there any workaround to prevent getting the "bad_weak_pointer" exception.

Reply:
Oh I see. In that case what I would do is have a constructor with a raw pointer, `GLMCostFunction(GLM*)`, and store that in `GLMCostFunction`. And to make sure you don't have some weird behaviour later on, I would replace `std::shared_ptr<GLMCostFunction> m_cost_function;` with `std::unique_ptr<GLMCostFunction> m_cost_function;`, so now you know there is only one reference to the pointer that points to `GLM`, so you won't have a memory leak or segfault when `GLM` is destructed :)